Improved DRMAAJobRunner

bernt-matthias opened the pull request:
A subclass of the DRMAAJobRunner that allows to detect time and memory violations. I took some inspiration from the slurm runner. I would be happy if I would get some comments, in particular if the logic of the main function will integrate nicely in the way jobs are pipelined through the runner. 

The main imrovements over the current DRMAAJobRunner are: 
- memory violations are determined by checking the stderr output and by comparing the used and the requested memory
- run time violations are determined by checking the signal that killed the job and by comparing the used and the requested run time
Where the used memory and time are determined with drmaa.wait() or qacct

In addition it solves the problem that the current DRMAAJobRunner does not work when jobs are submitted as real user (because jobs that are started in a different drmaa session can not be accessed from the session that is open in galaxy): 
- this is done by resorting to command line tools qstat and qacct if the drmaa library can not be used to check the job status and to get run time information. 
- this has the additional advantage that if the drmaa library functions are not working (DRMAAJobRunner had implemented a repeated checking to handle this problem) the runner can still use the command line tools.

I tested it successfully for our Univa grid engine. Also job resubmission seems to work nicely.

Open: 
- adaptions to other grid engines. the current implementation is specific for the Univa grid engine. For other grid engines the output of the command line tools might be different
- is it a problem that jobs are started with drmaa library (external runner) and the checks are done with the command line? Probably not.
- there is probably more :)

commit cca871edfe4778444b212d2ebc07f3e50ef68e82 changes in addition that the job template is loaded in the external runner before the user is changed. otherwise the cluster_files_directory needs to be readable by all users.

bernt-matthias committed:
UFZ specific message for logging in to the galaxy

bernt-matthias committed:
UFZ specific message for logging in to the galaxy

bernt-matthias committed:
UFZ specific message for logging in to the galaxy

bernt-matthias committed:
UFZ specific message for logging in to the galaxy

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added drmaa.wait

bernt-matthias committed:
added a comment on a potential flaw in the slurm runner implementation

bernt-matthias committed:
added a comment on a potential flaw in the slurm runner implementation

bernt-matthias committed:
added initial implementation of a runner for Univa grid engine

bernt-matthias committed:
added initial implementation of a runner for Univa grid engine

bernt-matthias committed:
added initial implementation of a runner for Univa grid engine

bernt-matthias committed:
added initial implementation of a runner for Univa grid engine

bernt-matthias committed:
added a comment to _complete_terminal_job

bernt-matthias committed:
added a comment to _complete_terminal_job

bernt-matthias committed:
added a comment to _complete_terminal_job

bernt-matthias committed:
added a comment to _complete_terminal_job

bernt-matthias committed:
added small comment

bernt-matthias committed:
added small comment

bernt-matthias committed:
added small comment

bernt-matthias committed:
added small comment

bernt-matthias committed:
undone changes applied to the wrong branch

bernt-matthias committed:
undone changes applied to the wrong branch

bernt-matthias committed:
undone changes applied to the wrong branch

bernt-matthias committed:
undone changes applied to the wrong branch

Commit:
making travis a bit more happy

Commit:
making travis a bit more happy

Commit:
making travis a bit more happy

Commit:
making travis a bit more happy

Commit:
more for travis' happiness

Commit:
more for travis' happiness

Commit:
more for travis' happiness

Commit:
more for travis' happiness

Commit:
even more happiness

Commit:
even more happiness

Commit:
even more happiness

Commit:
even more happiness

bernt-matthias committed:
made state a string (instead of set)

bernt-matthias committed:
made state a string (instead of set)

bernt-matthias committed:
made state a string (instead of set)

bernt-matthias committed:
made state a string (instead of set)

bernt-matthias committed:
output of subprocess is string not file -> change iteration

bernt-matthias committed:
output of subprocess is string not file -> change iteration

bernt-matthias committed:
output of subprocess is string not file -> change iteration

bernt-matthias committed:
output of subprocess is string not file -> change iteration

bernt-matthias committed:
Merge branch 'drmaa-fix' of https://github.com/bernt-matthias/galaxy into drmaa-fix

bernt-matthias committed:
added a new function check_watched_item
- the content of the function is the part of the original check_watched_items function
  that determined the job state with drmaa.job_state + exception handling
- the check_watched_items now just calls the check_watched_item function
The rationale for this change is that the drmaa.job_state function does not
work if jobs are submitted as real user. With the new function the method
to determine the job state can be changed in sub classes.

bernt-matthias committed:
added a new function check_watched_item
- the content of the function is the part of the original check_watched_items function
  that determined the job state with drmaa.job_state + exception handling
- the check_watched_items now just calls the check_watched_item function
The rationale for this change is that the drmaa.job_state function does not
work if jobs are submitted as real user. With the new function the method
to determine the job state can be changed in sub classes.

bernt-matthias committed:
added a new function check_watched_item
- the content of the function is the part of the original check_watched_items function
  that determined the job state with drmaa.job_state + exception handling
- the check_watched_items now just calls the check_watched_item function
The rationale for this change is that the drmaa.job_state function does not
work if jobs are submitted as real user. With the new function the method
to determine the job state can be changed in sub classes.

bernt-matthias committed:
added a new function check_watched_item
- the content of the function is the part of the original check_watched_items function
  that determined the job state with drmaa.job_state + exception handling
- the check_watched_items now just calls the check_watched_item function
The rationale for this change is that the drmaa.job_state function does not
work if jobs are submitted as real user. With the new function the method
to determine the job state can be changed in sub classes.

bernt-matthias committed:
version that survived the first tests

bernt-matthias committed:
version that survived the first tests

bernt-matthias committed:
version that survived the first tests

bernt-matthias committed:
version that survived the first tests

Commit:
check_watched_item now uses only job_info/qstat
this solves the problem that the wait called from _complete_terminal_job was unsuccessful (which led to the use of qacct) because it can only be called once for completed jobs (but this was already done in last call from check_watched_item)

Commit:
check_watched_item now uses only job_info/qstat
this solves the problem that the wait called from _complete_terminal_job was unsuccessful (which led to the use of qacct) because it can only be called once for completed jobs (but this was already done in last call from check_watched_item)

Commit:
check_watched_item now uses only job_info/qstat
this solves the problem that the wait called from _complete_terminal_job was unsuccessful (which led to the use of qacct) because it can only be called once for completed jobs (but this was already done in last call from check_watched_item)

Commit:
check_watched_item now uses only job_info/qstat
this solves the problem that the wait called from _complete_terminal_job was unsuccessful (which led to the use of qacct) because it can only be called once for completed jobs (but this was already done in last call from check_watched_item)

Commit:
minor bug fix

Commit:
minor bug fix

Commit:
minor bug fix

Commit:
minor bug fix

bernt-matthias committed:
bug fixes in time/memory parsing

bernt-matthias committed:
bug fixes in time/memory parsing

bernt-matthias committed:
bug fixes in time/memory parsing

bernt-matthias committed:
bug fixes in time/memory parsing

bernt-matthias committed:
making travis happy again

bernt-matthias committed:
making travis happy again

bernt-matthias committed:
making travis happy again

bernt-matthias committed:
making travis happy again

bernt-matthias committed:
bugfixes in time parsing

bernt-matthias committed:
bugfixes in time parsing

bernt-matthias committed:
bugfixes in time parsing

bernt-matthias committed:
bugfixes in time parsing

bernt-matthias committed:
Merge branch 'drmaa-fix' of https://github.com/bernt-matthias/galaxy into drmaa-fix

bernt-matthias committed:
reclaim ownership also for job restart, solves #4308

bernt-matthias committed:
reclaim ownership also for job restart, solves #4308

bernt-matthias committed:
reclaim ownership also for job restart, solves #4308

bernt-matthias committed:
reclaim ownership also for job restart, solves #4308

bernt-matthias commented:
Commit solves #4308. But I do not know if this is the best place.

mvdbeek committed:
Merge remote-tracking branch 'origin/release_17.05'

bernt-matthias committed:
improved run time and memory parsing

bernt-matthias committed:
improved run time and memory parsing

bernt-matthias committed:
improved run time and memory parsing

bernt-matthias committed:
improved run time and memory parsing

bernt-matthias committed:
added repeated polling of qacct (since accounting data might not be available immediately)

bernt-matthias committed:
added repeated polling of qacct (since accounting data might not be available immediately)

bernt-matthias committed:
added repeated polling of qacct (since accounting data might not be available immediately)

bernt-matthias committed:
added repeated polling of qacct (since accounting data might not be available immediately)

bernt-matthias committed:
Merge branch 'drmaa-fix' of https://github.com/bernt-matthias/galaxy into drmaa-fix

bgruening commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+        if tme is None:
+            logging.error( "DRMAAUniva: job {job_id} has unparsable time native spec {spec}".format(job_id=job_id, spec=native_spec) )
+    # parse memory
+    m = re.search( "mem=([0-9.]+[KGM]?)[\s,]*", native_spec )
```
@bernt-matthias quick question, does Univa also have an `-xml` option format the output? This would make parsing way more stable than the regex searches. I know that a few SGE commands do support the `-xml` argument.

bernt-matthias replied:
Unfortunately no (at least the version that is installed on our cluster).

bernt-matthias committed:
external runner: load job template before switching to other user

bernt-matthias committed:
external runner: load job template before switching to other user

bernt-matthias committed:
external runner: load job template before switching to other user

bernt-matthias committed:
added debug

bernt-matthias committed:
added debug

bernt-matthias committed:
added debug

bernt-matthias committed:
Merge branch 'drmaa-fix' of https://github.com/bernt-matthias/galaxy into drmaa-fix

bernt-matthias committed:
Merge remote-tracking branch 'upstream/master' into dev

natefoo commented on lib/galaxy/jobs/runners/drmaa.py:
```diff
@@ -68,6 +68,9 @@ def __init__( self, app, nworkers, **kwargs ):
         from pulsar.managers.util.drmaa import DrmaaSessionFactory
 
+        # make the drmaa library also available to subclasses
+        self.drmaa = drmaa
```
Is it not possible to import drmaa in the subclasses?

bernt-matthias replied:
The following made me think that this is a good solution: 
The UnivaJobRunner calls the superclass (DrmaaJobRunner) init function which does the import. So importing again in UnivaJobRunner seemed not necessary. Also the init function of DRMAAJobRunner takes quite some effort to import drmaa see: line 56-68 (eg. setting/using library paths) and I wanted to reuse this .. Furthermore it could be that there is some naming confusion because there is a global drmaa variable in DRMAAJobRunner (but I can not remember 100%).

Maybe the real question is if the import should be done in the init function?

bernt-matthias replied:
In the same way self.drmaa_job_states on the following lines is kind of ugly which is used then in the SLURMRunner. I basically have drawn my inspiration from this.

natefoo replied:
Ah right, this is to support the `$DRMAA_LIBRARY_PATH` environment variable. Making the module itself an instance attribute seems a bit dirty to me, though, so I'm going to attempt to do this a different way and make a PR against your branch if I get something working that I like.

natefoo commented on lib/galaxy/jobs/runners/slurm.py:
```diff
@@ -111,6 +111,7 @@ def _get_slurm_state():
                     check_memory_limit_msg = self.__check_memory_limit( ajs.error_file )
                     if check_memory_limit_msg:
                         log.info( '(%s/%s) Job hit memory limit', ajs.job_wrapper.get_id_tag(), ajs.job_id )
+                        # this fail message is odd since check_memory_limit_msg is True or False
```
I believe it's a non-empty string or False

bernt-matthias replied:
You are right. I removed the comment and added documentation to __check_memory_limit.

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+    def check_watched_item( self, ajs, new_watched ):
+        # get state with job_info/qstat
+        state = self._get_drmaa_state(ajs.job_id, self.ds, False)
+        logging.debug("DRMAAUnivaJobRunner:check_watched_item ({jobid}) -> state {state}".format(jobid=ajs.job_id, state=self.drmaa_job_state_strings[state]))
```
These log statements should use the log object acquired at the top, i.e. `logging.debug` -> `log.debug`. Also, for consistency, you shouldn't need to preface with `DRMAAUnivaJobRunner` since the module name is included in the message.

bernt-matthias replied:
has been replaced in the whole file

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+            return self.drmaa.JobState.UNDETERMINED
+
+
+def _check_memory_limit( efile_path ):
```
Rather than copying this from `SlurmJobRunner`, it'd probably be better to move it to somewhere like `galaxy.util`, remove the references to memory, and call it `grep_tail` or similar, and just pass in the set of strings to look for.

bernt-matthias replied:
good idea: created `grep_tail` in `galaxy.util` and adapted the `SLURMRunner` accordingly.

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+        logging.debug("DRMAAUnivaJobRunner._get_drmaa_state_qacct ({jobid}) -> {state}".format(jobid=job_id, state=self.drmaa_job_state_strings[state]))
+        return state
+
+    def _get_drmaa_state_wait( self, job_id, ds, extinfo ):
```
This could move up to the parent class since it's DRM-agnostic.

bernt-matthias replied:
Could be done, but does this make sense if no other runner uses it? What would be the rationale?

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+    pass
+
+
+class DRMAAUnivaJobRunner( DRMAAJobRunner ):
```
For consistency with the `SlurmJobRunner` runner, I'd go with just `UnivaJobRunner`.

bernt-matthias replied:
has been replaced in the whole file

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+
+
+class DRMAAUnivaJobRunner( DRMAAJobRunner ):
+    runner_name = "DRMAAUnivaJobRunner"
```
And `UnivaRunner`

bernt-matthias replied:
has been replaced in the whole file

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+            logging.debug("DRMAAUnivaJobRunner:_complete_terminal_job ({jobid}) memviolation {mv}".format(jobid=ajs.job_id, mv=memviolation))
+
+            # check job for run time or memory violation
+            if "deleted" in extinfo and extinfo[ "deleted" ]:
```
Whitespace, should be `if "deleted" in extinfo and extinfo["deleted"]:`

bernt-matthias replied:
done

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+            self.drmaa.JobState.USER_SUSPENDED: 0x22,
+            self.drmaa.JobState.DONE: 0x30,
+            self.drmaa.JobState.FAILED: 0x40 }
+        return drmaa_job_state_order[ staten ] > drmaa_job_state_order[ statep ]
```
Spaces around item access here as well.

bernt-matthias replied:
done

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+        """
+        # definition of the severity of job states, the hex codes are
+        # from the drmaa C library which seem to define a useful order
+        drmaa_job_state_order = {
```
Can this be merged with `drmaa_job_state_strings` in the base DRMAA runner? Perhaps the values should be another dict with keys `code` and `message`? Or if you wanted to be really fancy, make it a class and define the comparison operator methods so you can then just compare instances of the class.

bernt-matthias replied:
Correct, the cleanest solution would be a class. But then I would need to transform all job states produced by any DRMAARunner into JobState objects .. seems quite a bit of work and I can only test the standard DRMAA runner and the Univa runner. 

Is it worth the effort? Or would you rather prefer the dict based approach that you have suggested?

natefoo commented on lib/galaxy/jobs/runners/drmaauniva.py:
```diff
@@ -0,0 +1,640 @@
+        # be more efficient to specify the user (or in case that the galaxy user
+        # submits the job -> to ommit the -u option)
+        cmd = ['qstat', '-u', '"*"']
+        p = subprocess.Popen( cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE )
```
To avoid duplication these should probably be reduced to a single utility function (I should do this in the Slurm runner as well!). Maybe `check_output_error` mimicking the behavior of `subprocess.check_output` (which there's a copy of [here](https://github.com/galaxyproject/starforge/blob/f1013ddf6119381ffba6df06e8d996cee5678280/starforge/util.py#L43)), just return the stderr with it as a tuple, e.g. `return (output, error)`

bernt-matthias replied:
Hmm. Got the idea and started coding, but was **not entirely convinced** that it is useful since the code in my python class gets larger... 

The reason is that I often check stderr/errorcode quite a bit. So I would need a function that returns stdout, stderr, and the errorcode. Furthermore the function can not raise an exception if stderr!="" or errorcode because I still need these outside of the function. 

Hence I would only replace two lines (Popen and communicate) by one (the new function call).

natefoo reviewed (changes requested):
Thanks for working on this @bernt-matthias! I haven't been able to review this quite as thoroughly as I wanted so I might have a few more comments later, but I added what I could so far.

bernt-matthias committed:
logging -> log
DRMAAUnivaJobRunner -> UnivaJobRunner

bernt-matthias committed:
logging -> log
DRMAAUnivaJobRunner -> UnivaJobRunner

bernt-matthias committed:
logging -> log
DRMAAUnivaJobRunner -> UnivaJobRunner

bernt-matthias committed:
logging -> log
DRMAAUnivaJobRunner -> UnivaJobRunner

bernt-matthias committed:
try to use memory parsing from util

bernt-matthias committed:
try to use memory parsing from util

bernt-matthias committed:
try to use memory parsing from util

bernt-matthias committed:
try to use memory parsing from util

bernt-matthias committed:
use utils memory parsing at all places

bernt-matthias committed:
use utils memory parsing at all places

bernt-matthias committed:
use utils memory parsing at all places

bernt-matthias committed:
use utils memory parsing at all places

bernt-matthias committed:
created a common function in util that checks the tail of a file
for substrings

bernt-matthias committed:
created a common function in util that checks the tail of a file
for substrings

bernt-matthias committed:
created a common function in util that checks the tail of a file
for substrings

bernt-matthias committed:
created a common function in util that checks the tail of a file
for substrings

bernt-matthias commented:
I fixed most of the comments of @natefoo or raised questions. Thanks a lot for the input.

bernt-matthias committed:
util.size_to_bytes: accept also memory strings without multiplier K/M/G/T

bernt-matthias committed:
util.size_to_bytes: accept also memory strings without multiplier K/M/G/T

bernt-matthias committed:
util.size_to_bytes: accept also memory strings without multiplier K/M/G/T

bernt-matthias committed:
util.size_to_bytes: accept also memory strings without multiplier K/M/G/T

bernt-matthias commented:
the change in util.size_to_bytes needs more thorough checking (it now also accepts strings without a multiplier K/M/G/T), because of possible side effects at other places

natefoo commented:
Thanks for all the changes! I'm going to take another look through and answer any remaining questions you raised shortly.

bernt-matthias committed:
Merge remote-tracking branch 'upstream/dev' into dev

bernt-matthias committed:
missed something while merging

bernt-matthias committed:
missed something while merging

bernt-matthias committed:
Merge branch 'drmaa-fix' of https://github.com/bernt-matthias/galaxy into drmaa-fix

bernt-matthias commented on scripts/drmaa_external_runner.py:
```diff
@@ -116,15 +121,20 @@ def set_user(uid, assign_all_groups):
 def main():
     userid, json_filename, assign_all_groups = validate_paramters()
-    set_user(userid, assign_all_groups)
+    # load JSON job template data before changing the user
```
The switch in the order of the commands is not strictly necessary (I was just playing around -- but the problem that I was trying to solve was wrong permissions in the file system), but it will simplify my next PR. This will be to use `sudo -u ` instead of `sudo`

bernt-matthias commented:
I think I messed up the merging with dev the changes in static/ are not intended. Still need to learn a lot with git .. sorry

mvdbeek committed:
Fix startup of galaxy when wehbooks dir is empty

Reported by @alpapan on gitter.

hexylena committed:
Merge pull request #4570 from mvdbeek/webhook_loading_fix

[17.05] Bug: Fix startup of galaxy when wehbooks dir is empty

dannon committed:
Allow fetchMore to actually specify a 0 offset when intentional.  This should fix the mysterious missing history in multiview bug.

jmchilton committed:
Merge pull request #4615 from dannon/release_16.10_backport2

[16.10] Allow fetchMore to actually specify a 0 offset when intentional, fixes multiview

dannon committed:
Merge branch 'release_16.10' into release_17.01

nsoranzo committed:
Fix virtualenv activation for some scripts

In particular add virtualenv activation to extract_dataset_parts.sh
to fix the following error when running tools which use
`<parallelism>`:
```
Fatal error:
/tmp/tmpOoQ01k/job_working_directory/000/12/task_0:
Traceback (most recent call last):
File "./scripts/extract_dataset_part.py", line 17, in <module>
import galaxy.model.mapping  # need to load this before we unpickle, in order to setup properties assigned by the mappers
File "/opt/galaxyproject_galaxy_release_17.05/lib/galaxy/model/__init__.py", line 20, in <module>
from six import string_types
ModuleNotFoundError: No module named 'six'
```

dannon committed:
Merge branch 'release_17.01' into release_17.05

dannon committed:
Merge pull request #4624 from nsoranzo/release_17.05_fix_scripts

[17.05] Backport of #4616

nsoranzo committed:
Merge branch 'release_17.05'

bernt-matthias commented:
Since I had problems to perform a tutorial with 12 users I did a stress test (submitting ~50 1second jobs at once) and debugging with the runner. 

The problem was that `qacct` which is called with `Popen` needs extremely long on our system: ~3 min. It appeared to me that the call to Popen does not release python's GIL. I get the same results independent of the number of workers. 

I have solved this for our system by replacing the installed qacct by a self developed python script which greps the accounting DB from the back... :) Are there experiences how fast 50 concurrent jobs should need (I'm still using a setup with a single galaxy process)? 

Still the Popen question bugs me since it is also used in other places: the external runners. Even if I guess this wont be a big problem as long as they are fast... But there might be circumnstances (slow cluster file system performance) where this might become a problem.

bernt-matthias committed:
removed/commented log output + added comments

bernt-matthias committed:
removed/commented log output + added comments

bernt-matthias committed:
Merge branch 'dev' of https://github.com/bernt-matthias/galaxy into dev

bernt-matthias committed:
Merge remote-tracking branch 'upstream/master' into dev

bernt-matthias committed:
Merge branch 'drmaa-fix' of https://github.com/bernt-matthias/galaxy into drmaa-fix

nsoranzo commented:
Superseded by #4857.
